### PR TITLE
Improve dynamics model training

### PR DESCRIPTION
**Pull request by blake-wulfe-tri**
This commit improves dynamics model training by:
1. Adding normalization of model inputs / outputs
2. Logging visualizations of predictions during training
3. Changing the action distribution during data generation to match that used during evaluation

Given those changes, planning using random shooting with a learned model consistently outperforms planning with a random model, and seems to "work".

- - -
This change is [Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23)

### CODE REVIEW

**Review by blake-wulfe-tri — COMMENTED** [review]
+@hongkai-dai for review please

w/ trained model:


w/o trained model:


tensorboard images:


> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23#-:-NPaAQaME_MMl9XtNFWk:b1wq27h)* status: 0 of 15 files reviewed, all discussions resolved (waiting on @hongkai-dai)

**Review by hongkai-dai — CHANGES_REQUESTED** [review]
> Reviewed 14 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23)* status: 14 of 15 files reviewed, 1 unresolved discussion (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dynamics/modules.py` line 63 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23#-NPfIqy138kXvwmoCxQc:-NPfIqy138kXvwmoCxQd:bbuvm2b) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/7c2e4be6d5c773cc770f9be729144ae2db04f88c/mani_skill2/dynamics/modules.py#L63)):*
> ```Python
>         # Only normalize next_state and pred_next_state for the loss calculation
>         # if the pred_next_state has already been denormalized.
>         if self.dynamics_model.output_normalizer is not None:
> ```

Could you talk more why we need to normalize the output? Why not to embed this normalization into the `self._compute_loss` function?

**Review by blake-wulfe-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23)* status: 14 of 15 files reviewed, 1 unresolved discussion (waiting on @hongkai-dai)

___
*[`mani_skill2/dynamics/modules.py` line 63 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23#-NPfIqy138kXvwmoCxQc:-NPrlzOy0G3AaLfk5cnY:bhhy9ss) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/7c2e4be6d5c773cc770f9be729144ae2db04f88c/mani_skill2/dynamics/modules.py#L63)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Could you talk more why we need to normalize the output? Why not to embed this normalization into the `self._compute_loss` function?
</blockquote></details>

hmm maybe, let me think through it
- The point is to make the optimization for the network easier. 
- The problem is that if you don't normalize the output, then the model has to learn to have very small output values (e.g., 0.01). Given typical learning rates and weight initializations that is difficult to do / slows down learning.
- To fix that, I have the model output values 10x larger, then scale them down 10x when returning them from the dynamics model class. This I am confident is necessary based on prior experiments.
- Then in the loss, I scale the next state and pred next state by 10x to compute the loss. But why do I do that? Is that actually necessary? I think all it's doing is trading off the weight on the l2 norm vs l1 norm, and otherwise scaling the loss. The scaling of the loss is basically the same as scaling the learning rate. So my guess is that this loss scaling isn't even necessary. Let me check (removing the loss scaling and increasing the learning rate 10x):
w/ loss scaling




w/o loss scaling





Quantitatively, the planning performance is much better with the loss scaled approach, but having said that, based on the logic above, it's likely that there's some hyperparameter configuration that would give equivalent results (e.g., the difference in performance is b/c training w/ learning rate of 1e-3 produces a worse model). So I'll remove the scaling for now.

**Review by hongkai-dai — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23)* status: 12 of 15 files reviewed, 2 unresolved discussions (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dynamics/network_dynamics_model.py` line 57 at r2](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23#-NPsFeG0C1_32GQLAvrm:-NPsFeG16WyFKRg_b7up:b2g75kj) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/2acaa29d5806b2f68a2647f31085beff95368b3c/mani_skill2/dynamics/network_dynamics_model.py#L57)):*
> ```Python
>             state = self.state_normalizer.normalize(state)
>         if self.action_normalizer is not None:
>             action = self.action_normalizer.normalize(action)
> ```

The CI failure seems real. It is `act` instead of `action` in the function input.

**Review by blake-wulfe-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23)* status: 11 of 16 files reviewed, 2 unresolved discussions (waiting on @hongkai-dai)

___
*[`mani_skill2/dynamics/network_dynamics_model.py` line 57 at r2](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23#-NPsFeG0C1_32GQLAvrm:-NPsIQJIEb6CmIApAfJc:bkxwz0h) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/2acaa29d5806b2f68a2647f31085beff95368b3c/mani_skill2/dynamics/network_dynamics_model.py#L57)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

The CI failure seems real. It is `act` instead of `action` in the function input.
</blockquote></details>

thanks for looking into that

**Review by hongkai-dai — CHANGES_REQUESTED** [review]
:lgtm: sorry for the delay.

> Reviewed 2 of 2 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23#-:-NPthJoR8O8Wt2kudSnF:b-c82x7h)* status: all files reviewed (commit messages unreviewed), 1 unresolved discussion (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dynamics/modules.py` line 63 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23#-NPfIqy138kXvwmoCxQc:-NPtgXUKEdOYgKYcb81V:b9ibsmo) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/7c2e4be6d5c773cc770f9be729144ae2db04f88c/mani_skill2/dynamics/modules.py#L63)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

hmm maybe, let me think through it
- The point is to make the optimization for the network easier. 
- The problem is that if you don't normalize the output, then the model has to learn to have very small output values (e.g., 0.01). Given typical learning rates and weight initializations that is difficult to do / slows down learning.
- To fix that, I have the model output values 10x larger, then scale them down 10x when returning them from the dynamics model class. This I am confident is necessary based on prior experiments.
- Then in the loss, I scale the next state and pred next state by 10x to compute the loss. But why do I do that? Is that actually necessary? I think all it's doing is trading off the weight on the l2 norm vs l1 norm, and otherwise scaling the loss. The scaling of the loss is basically the same as scaling the learning rate. So my guess is that this loss scaling isn't even necessary. Let me check (removing the loss scaling and increasing the learning rate 10x):
w/ loss scaling




w/o loss scaling





Quantitatively, the planning performance is much better with the loss scaled approach, but having said that, based on the logic above, it's likely that there's some hyperparameter configuration that would give equivalent results (e.g., the difference in performance is b/c training w/ learning rate of 1e-3 produces a worse model). So I'll remove the scaling for now.
</blockquote></details>

Thanks a lot for the experiments and the explanation!

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/23)* status: all files reviewed, 1 unresolved discussion (waiting on @blake-wulfe-tri)
